processor-api/src/main/java/com/synopsys/integration/alert/processor/api/summarize/DefaultProjectMessageSummarizer.java

```diff
@@ -0,0 +1,174 @@
+                    "Alert received an unrecognized notification from BlackDuck."
+                );
+        }
+    }
+
+    private Pair<String, String> constructProjectVersionStatusSummaryAndDescription(String providerName, String projectName, ProjectMessage projectMessage) {
+        String operationString = projectMessage.getOperation()
+                                     .map(this::convertToParticiple)
+                                     .orElse("updated");
+
+        Optional<String> optionalProjectVersionName = projectMessage.getProjectVersion().map(LinkableItem::getValue);
+        if (optionalProjectVersionName.isPresent()) {
+            String projectVersionName = optionalProjectVersionName.get();
+            return Pair.of(
+                String.format("[%s] %s > %s %s", providerName, projectName, projectVersionName, operationString),
+                String.format("A project-version '%s' was %s was in the BlackDuck project '%s'.", projectVersionName, operationString, projectName)
+            );
+        } else {
+            return Pair.of(
+                String.format("[%s] %s %s", providerName, projectName, operationString),
+                String.format("The project '%s' was %s in BlackDuck.", projectName, operationString)
+            );
+        }
+    }
+
+    private Pair<String, String> constructComponentStatusAndDescription(String providerName, String projectName, MessageReason messageReason, ProjectMessage projectMessage) {
+        String projectVersionName = projectMessage.getProjectVersion()
+                                        .map(LinkableItem::getValue)
+                                        .orElse("Unknown Version");
+        String operationString = MessageReason.COMPONENT_CONCERN.equals(messageReason) ? "problems" : "updates";
+        return Pair.of(
+            String.format("[%s] %s > %s component %s", providerName, projectName, projectVersionName, operationString),
+            String.format("The project-version '%s > %s' had component %s in BlackDuck", projectName, projectVersionName, operationString)
+        );
+    }
+
+    private List<LinkableItem> constructMessageDetails(ProjectMessage projectMessage) {
+        List<LinkableItem> details = new LinkedList<>();
+        details.add(projectMessage.getProvider());
+        details.add(projectMessage.getProject());
+        projectMessage.getProjectVersion().ifPresent(details::add);
+
+        Map<Triple<ComponentConcernType, ItemOperation, ComponentConcernSeverity>, Integer> concernCounts = new LinkedHashMap<>();
+        for (BomComponentDetails bomComponent : projectMessage.getBomComponents()) {
+            for (ComponentConcern concern : bomComponent.getComponentConcerns()) {
+                Triple<ComponentConcernType, ItemOperation, ComponentConcernSeverity> concernKey = Triple.of(concern.getType(), concern.getOperation(), concern.getSeverity());
```

Comment:
Does it make sense to have a data class here? I find the `Triple` makes the code more difficult to read. It seems rather contained though.

Reply:
I was going back and forth on it since the data class would not be exposed as part of any API, but I think you're right that readability is somewhat degraded.